hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/OmBucketReadWriteKeyOps.java

```diff
@@ -128,12 +128,10 @@ protected String createPath(String path) {
   @Override
   protected int getReadCount(int readCount, String readPath)
       throws IOException {
-    Iterator<? extends OzoneKey> ozoneKeyIterator = bucket.listKeys(
-        OzoneConsts.OM_KEY_PREFIX + readPath + OzoneConsts.OM_KEY_PREFIX);
-    while (ozoneKeyIterator.hasNext()) {
-      ozoneKeyIterator.next();
-      ++readCount;
-    }
+    List<OzoneFileStatus> ozoneFileStatusList = bucket.listStatus(
+        OzoneConsts.OM_KEY_PREFIX + readPath + OzoneConsts.OM_KEY_PREFIX, true,
+        "/", keyCountForRead);
```

Comment:
If ``keyCountForRead`` is huge, in that case does it go and try to fetch all entries in one go and store it in ``List<OzoneFileStatus> ozoneFileStatusList``, in that case for larger values of ``keyCountForRead``, this might start giving OOM.

Reply:
Ideally the server should limit the max number of entries returned. This is a CLI side change that should not be implementing such caps.

Reply:
On a quick look I think here client is controlling the number of entries via ``keyCountForRead``.

In the FileSystem code as well, it seems to be controlled at the client side:
https://github.com/apache/ozone/blob/a32549cbc1227291d6ebd972ce32db043a5c51d0/hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneFileSystem.java#L608

In HDFS, AFAIK the Namenode does impose such limits and the client can't explicitly demand the number of entries:
https://github.com/apache/hadoop/blob/123d1aa8846a2099b04fd8e8ed7f2bd6db4f36b5/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirectory.java#L364-L366

And the use case here seems to be just getting the ``readCount``, which actually doesn't require ``OzoneFileStatus``, it can easily be looped using the ``LISTING_PAGE_SIZE`` or similar numbers, If you go at a scale of around a ~1 Million entries or beyond you would be utilising a lot of memory storing the ``FileStatus`` which you actually don't need, If I remember correct it started taking around ~1.5 Gigs/million, when I tried similar stuff with HDFSFileStatus in DistCp which sadly builds up the FileStatus tree in Breadth First Traversal approach. It would be less for Ozone for sure as compared to HDFS....

I couldn't spare time to dig in much, Please feel free to proceed if you are convinced.

Reply:
@ayushtkn what you point out makes sense. The reason it is not relevant here is that freon is a CLI extension used to generate load for testing. The user might not want to have a limit as they might be profiling memory or CPU. The CLI invocation has a [default value that can be overridden ](https://github.com/apache/ozone/pull/3554/files#diff-1fa33049586e9221ab06c5c11fba5033a8ca68a28b8fe5ccefe4b676ef4f9c0fR62-R65)

Reply:
> @ayushtkn what you point out makes sense. The reason it is not relevant here is that freon is a CLI extension used to generate load for testing. The user might not want to have a limit as they might be profiling memory or CPU. The CLI invocation has a [default value that can be overridden ](https://github.com/apache/ozone/pull/3554/files#diff-1fa33049586e9221ab06c5c11fba5033a8ca68a28b8fe5ccefe4b676ef4f9c0fR62-R65)

Thanks a lot @ayushtkn for the reviews. 

Agreed with @kerneltime. Since its a stress/benchmarking tool, I am more inclined to keep this value open to the users and they can tune this based on their cluster resource capabilities.